File: preproc/preproc.f
      PROGRAM PREPROC_GAUGE
C ******************************************************************
C * THIS PROGRAM READS RAW RAIN GAUGE DATA FROM GV FIELD SITES AND 
C * FORMATS DATA TO BE COMPATIBLE WITH RGMIN.F.  FILENAMES ARE NAMED U
C * SING CONVENTION ESTABLISHED FOR RGMIN INPUT DATA.  MAIN PROGRAM 
C * CONSISTS OF CALLS TO SITE MODULES WHICH DO THE ACTUAL PROCESSING 
C * AND WRITING OF DATA. THE PROGRAM REQUIRES FOUR COMMAND LINE ARGUMENTS.
C *        1) Top level Directory path 
C *        2) NETWORK ID - 3 CHARACTER CODE ASSOCIATING GAUGE W/NETWORK
C *        3) BYEAR - beginning year of data
C *        4  VERBOSE - Verbose Flag
C*****************************************************************************

      CHARACTER*60 FILELIST
      CHARACTER*3  NETWORK_ID
      CHARACTER*2  BYEAR
      CHARACTER*1  VERBOSE
C
C *** Check to see that the number of command line arguments is correct.
C
      NARG = IARGC()
      IF(NARG.NE.4) THEN 
         WRITE(6,*)'Usage:'
         WRITE(6,*)'preproc DIRECTORY SITEID YEAR VERBOSE'
         WRITE(6,*)'DIRECTORY: above the data/ directory'
         WRITE(6,*)'SITEID is a 3 character site identifier'
         WRITE(6,*)'YEAR beginning year of data'
         WRITE(6,*)'VERBOSE is the verbose flag (1 = verbose)'
         STOP
      ENDIF
C
C *** Get command line arguments
C
      CALL GETARG(1,FILELIST)
      CALL GETARG(2,NETWORK_ID)
      CALL GETARG(3,BYEAR)
      CALL GETARG(4,VERBOSE)
      READ(VERBOSE,10)IVERBOSE
 10   FORMAT(I1)

         WRITE(6,'("FILELIST: ",A60)') FILELIST
         WRITE(6,'("NETWORK_ID: ",A6)') NETWORK_ID
         WRITE(6,'("BYEAR: ",A2)') BYEAR
         WRITE(6,'("VERBOSE: ",A1)') VERBOSE
C
C *** Call preprocessing module based on NETWORK_ID given at command line
C
      IF(NETWORK_ID.EQ.'STJ') THEN
         WRITE(6,*) 'PROCESSING DATA FROM St. Johns WMD'
         CALL PREPROC_STJ(NETWORK_ID,FILELIST,BYEAR,IVERBOSE)
      ELSEIF(NETWORK_ID.EQ.'RMI'.OR.NETWORK_ID.EQ.'KWA') THEN
         WRITE(6,*) 'PROCESSING DATA FROM the Marshall Islands'
         CALL PREPROC_KWA(NETWORK_ID,FILELIST,BYEAR,IVERBOSE)
      ELSEIF(NETWORK_ID.EQ.'CSC'.OR.NETWORK_ID.EQ.'DSC'.OR.
     +        NETWORK_ID.EQ.'MSC') THEN
         WRITE(6,*) 'PROCESSING DATA FROM the Darwin, Australia'
         CALL PREPROC_DAR(NETWORK_ID,FILELIST,BYEAR,IVERBOSE)
      ELSEIF(NETWORK_ID.EQ.'TAM') THEN
         WRITE(6,*) 'PROCESSING DATA FROM Texas A&M'
         CALL PREPROC_TAM(NETWORK_ID,FILELIST,BYEAR,IVERBOSE)
      ELSEIF(NETWORK_ID.EQ.'KSC') THEN
         WRITE(6,*) 'PROCESSING DATA FROM KSC'
         CALL PREPROC_KSC(NETWORK_ID,FILELIST,BYEAR,IVERBOSE)
      ELSEIF(NETWORK_ID.EQ.'HAR') THEN
         WRITE(6,*) 'PROCESSING DATA FROM HAR'
         CALL PREPROC_HAR(NETWORK_ID,FILELIST,BYEAR,IVERBOSE)
      ELSEIF(NETWORK_ID.EQ.'OMK') THEN
         WRITE(6,*) 'PROCESSING DATA FROM Om Koi, Thailand'
         CALL PREPROC_GEMS(NETWORK_ID,FILELIST,BYEAR,IVERBOSE)
      ELSEIF(NETWORK_ID.EQ.'KSO') THEN
         NETWORK_ID = 'KSC'
         WRITE(6,*) 'PROCESSING DATA FROM KSC (OLD FORMAT)'
         CALL PREPROC_GEMS(NETWORK_ID,FILELIST,BYEAR,IVERBOSE)
      ELSEIF(NETWORK_ID.EQ.'KWO')THEN
         WRITE(6,*) 'PROCESSING DATA FROM Kwajalein (OLD FORMAT)'
         CALL PREPROC_GEMS(NETWORK_ID,FILELIST,BYEAR,IVERBOSE)
      ELSEIF(NETWORK_ID.EQ.'SFL')THEN
         WRITE(6,*) 'PROCESSING DATA FROM South Florida WMD'
         CALL PREPROC_SFL(NETWORK_ID,FILELIST,BYEAR,IVERBOSE)
      ELSEIF(NETWORK_ID.EQ.'TFB')THEN
         WRITE(6,*) 'PROCESSING DATA FROM TFB'
         CALL PREPROC_KWA(NETWORK_ID,FILELIST,BYEAR,IVERBOSE)
      ELSEIF(NETWORK_ID.EQ.'LBA')THEN
         WRITE(6,*) 'PROCESSING DATA FROM Ji-Parana, Brazil'
         CALL PREPROC_KWA(NETWORK_ID, FILELIST,BYEAR,IVERBOSE)
      ELSEIF(NETWORK_ID.EQ.'TFT')THEN
         WRITE(6,*) 'PROCESSING DATA FROM TFT TOKAY'
         CALL PREPROC_TFT(NETWORK_ID,FILELIST,BYEAR,IVERBOSE)
      ENDIF
      END

C ********************************************************************
C *  GET_TIP BACKS OUT THE TIME OF EACH TIP AND RAINFALL AMOUNT BY   *
C * SUBTRACTING THE CURRENT RECORD FROM THE PREVIOUS RECORD.  THIS   *
C * PROCEDURE ALSO RESOLVES FOR MULTIPLE TIPS WHICH ARE ASSIGNED A   *
C * SINGLE TIME STAMP. CURRENTLY USED IN: HAR, STJ.                  *
C ********************************************************************
      SUBROUTINE GET_TIP(NR,RAIN,TIP)

      PARAMETER    (ADIM = 5000)
      REAL*4      RAIN(ADIM),TIP(NR)
      
      TIP(1) = 0
      DO I = 2,NR
         TIP(I) = RAIN(I) - RAIN(I-1)
         RAIN(I-1) = 0
      ENDDO
      RETURN
      END

C *******************************************************************
C *  This subroutine combines the NETWORK_ID, G_NUMBER, and CYR     *
C *  fields from which output filename is constructed.              *
C *******************************************************************
      SUBROUTINE NAMEFILE(NETWORK_ID,IFILE,G_NUMBER,CYR,OFILE)
      CHARACTER*60 IFILE,OFILE
      CHARACTER*4  G_NUMBER
      CHARACTER*3 NETWORK_ID
      CHARACTER*2 CYR
     
      OFILE = ''
C
C ***  Search directory path for position of "raw".  Path convention allows
C ***  output filename and directory path of output to be dynamically named.
C
      POS = INDEX(IFILE,'raw')


      OFILE = IFILE(1:POS-1)
      OFILE(POS:POS+3) = 'gag/'
      OFILE(POS+4:POS+6) = NETWORK_ID
      OFILE(POS+7:POS+7) = '/'
      OFILE(POS+8:POS+10) = NETWORK_ID
      OFILE(POS+11:POS+14) = G_NUMBER
      OFILE(POS+15:POS+15) = '_'
      OFILE(POS+16:POS+17) = CYR
      OFILE(POS+18:POS+21) = '.gag'
      
      RETURN
      END

File: preproc/preproc.pl
#!/usr/bin/perl
MAIN:
{
	$nargs = 4-1;
	if ($#ARGV < $nargs) {
        print STDERR "Usage:\n $0 DIRECTORY SITEID YEAR VERBOSE\n";
		print STDERR "DIRECTORY above the data/ directory\n";
		print STDERR "SITEID is arbitrary site identificatio\n";
		print STDERR "YEAR beginning year of data\n";
		print STDERR "VERBOSE is the verbose flag.\n";
        exit;
	}
#
# ### Set defaults
#
	$prog     = "preproc";		# Program name
	$verbose  = "0";			# Verbose = on
	$net_id   = "???";			# Network ID
#
# ### Assign command line parameter values
#
	$dir      = shift;
	$net_id   = shift;
	$year     = shift;
	$verbose  = shift;
	$in_dir   = $dir."/data/raw/".$net_id."/";    # Input directory
	$out_dir  = $dir."/data/gag/".$net_id."/";    # Output directory
	if($net_id eq "KSO"){
		$in_dir   = $dir."/data/raw/KSC/";    # Input directory
		$out_dir  = $dir."/data/gag/KSC/";    # Output directory#
	}

#
# ### Clean the output directories.
#
	if($verbose == 1) {
		print "Removing old GAG files from $out_dir\n";
	}
	$command = "rm $out_dir/*.gag";
	system($command);
#
# ### Print to screen
#
	if($verbose == 1) {
		printf("NETWORK_ID: %s\n", $net_id);
		printf("INPUT Directory: %s\n", $in_dir);
		printf("OUTPUT Directory: %s\n", $out_dir);
	}
#
# ### Get a list of all of the input (*.gag) files, write their
# ### names one per line in the $prog.i file.
#
	open(FILE,"> file.list"); 	# Create list of input filenames	
	open(LIST,"ls $in_dir |"); 	
	while(<LIST>) {
		chop;
		$i++;
		$raw_file = $in_dir."/".$_;
		$mod_file = $raw_file.".tmp";
		printf("RAWFILE == %s\n",$raw_file);
		printf("MODFILE == %s\n",$mod_file);
		#
		# Remove blank lines from the input file. We will 
		#
		open(RAWFILE,"cat $raw_file |");
		open(MODFILE,"> $mod_file");
		#
		# While not at EOF of raw file, and cat all non
		# blank lines to mod_file
		#
	 	while(<RAWFILE>) {
			chop;	# Remove \n from end of line
			$nlines++;
			#
			# *** If the line is not blank, write it out.
			#
			if($_ ne "") {
				#
				# *** Harris County data sometimes has an integer value for the
				# *** rainrate, especially if it is 0. This will crash har.f 
				# *** when it tries to read the data.
				# *** A similar problems occurs with the TAM data.
				#
				$reform = 0;
				
				#
				# *** Texas A&M Mesonet Data
				#
				if($net_id eq "TAM") {
					($f1, $f2, $f3, $f4, $f5, $f6) = split(",",$_);
					if($f1 eq "4") { # Tip record
						$f4 = $f4*1.;
						$f5 = $f5*1.;
						printf(MODFILE "%s,%s,%s,%.2f,%.2f\n", 
						 		  $f1, $f2, $f3, $f4, $f5);
					} else { # Normal record
						$f5 = $f5*1.;
						$f6 = $f6*1.;
						printf(MODFILE "%s,%s,%s,%s,%.2f,%.2f\n", 
						 		  $f1, $f2, $f3, $f4, $f5, $f6);
					}					
				    $reform = 1;
				}
				#
				# *** HARRIS COUNTY
				#
				if($net_id eq "HAR") {
					($date, $time, $rate) = split(" ",$_);
					printf(MODFILE "%s %s %-12.8f\n", 
						   $date, $time, $rate*1.0);
				    $reform = 1;
				} 
				if($reform == 0) {
					printf(MODFILE "$_\n");
				}

			}
		}
		close(RAWFILE);
		close(MODFILE);
		#
		# *** Copy the modified data over the raw data.
		#
		print("Copying modfile to raw file...\n");
		$command = "cp $mod_file $raw_file";
		system($command);

		$command = "rm $mod_file";
		system($command);
		print(FILE "$raw_file \n");
	}	
	close(LIST);			# 						
	close(FILE);
	$command = $prog." file.list ".$net_id." ".$year." ".$verbose;
	print $command."\n";
#
# ### Execute $prog using $prog.i in current directory.
#

	system($command);

}
